client/app/collections/realeventsgenerator.coffee

```diff
@@ -14,67 +14,157 @@ module.exports = class RealEventGeneratorCollection extends Backbone.Collection
         @listenTo @baseCollection, 'remove', @resetFromBase
         @listenTo @baseCollection, 'reset', @resetFromBase
 
-    resetFromBase: ->
-        @reset []
-        @trigger 'reset'
+        @listenTo app.calendars, 'change', -> @resetFromBase true
 
-    generateRealEvents: (start, end, callback) ->
+        @_initializeGenerator()
+
+    _initializeGenerator: ->
+        @previousRecurringEvents = []
+        @runningReccuringEvents = []
+
+        # Default value. In the particular case of all events > today,
+        # then last event is the closest to today.
+        @firstGeneratedEvent = @baseCollection.at @baseCollection.length - 1
+        @lastGeneratedEvent = null
+
+        today = moment().startOf 'day'
+        @firstDate = today.clone()
+        @lastDate = today.clone()
+
+        i = 0
+        while i < @baseCollection.length
+            item = @baseCollection.at i
+            i++
+
+            if not item.isVisible() then continue
+
+            # break loop condition
+            if item.getStartDateObject().isAfter today
+                @firstGeneratedEvent = item
+                @lastGeneratedEvent = item
+                break
+
+            if item.isRecurrent()
+                @previousRecurringEvents.push item
+
+                if item.getLastOccurenceDate().isAfter today
+                    @runningReccuringEvents.push item
+
+        @loadNextPage()
+
+    resetFromBase: (sync)->
+        resetProc = =>
+                @reset []
+                @_initializeGenerator()
+                @trigger 'reset'
+        if sync
+            resetProc()
+        else
+            # Asynchronous, to avoid lag on clicks and edits.
+            setTimeout resetProc, 1
+
+
+    loadNextPage: (callback) ->
         callback = callback or ->
+
         eventsInRange = []
-        @baseCollection.each (item) =>
+        start = @lastDate.clone()
+        @lastDate.add 1, 'month'
+        end = @lastDate.clone()
+
+        # pic ponctual event and store newly found recurring ones.
+        i = @baseCollection.indexOf @lastGeneratedEvent
+        while i >= 0 and i < @baseCollection.length
+            item = @baseCollection.at i
+            i++
 
-            calendar = item.getCalendar()
-            return null if calendar and calendar.get('visible') is false
+            if not item.isVisible() then continue
+
+            # End loop conditions
+            if item.getStartDateObject().isAfter end
+                @lastGeneratedEvent = item
+                break
 
             if item.isRecurrent()
-                evs = item.generateRecurrentInstancesBetween start, end, \
+                @runningReccuringEvents.push item
+
+            else
+                eventsInRange.push new RealEvent item
```

Comment:
The loop can be improved syntax-wise, by using the full power of a while loop and avoid the use of `break`

```coffeescript
lastEventIndex = @baseCollection.indexOf @lastGeneratedEvent
lastEventExist = lastEventIndex isnt -1 # using a variable to make the loop more semantic
@lastGeneratedEvent = null # I'm not sure it's necessary, but it might be
while lastEventIndexExist and i < @baseCollection.length and @lastGeneratedEvent is null
    item = @baseCollection. at i
    i++
    unless item.isVisible() then continue
    else if item.getStarteDateObject().isAfter(end) # always use parenthesis in conditional statements
        @lastGeneratedEvent = item
    else if item.isRecurrent()
        @runningRecurringEvents.push item
    else
        eventsInRange.push new RealEvent(item)
``
What do you think?